### ISSUE TITLE
Data Oriented Design

### ISSUE THREAD

**Issue by GPUCode** [PanosK92/SpartanEngine#26]
Hello, I recently discovered this awesome repository and I have been looking at the code for some time now. I am really fasinated by the architecture of the code and I have been wanting to contribute for some time. Having studied Data Oriented Design for some time now and looking at your entity class I was wondering if i could swap out that std::vectorof Components with ComponentManager classes which use an entity id to lookup components. In this way you can store components linearly in memory and also have functions like: GetAllTransformComponets(). I have impemented this design in my own engine/render with great results but mine is tiny compared to yours. Thanks for reading!

TLDR I am looking (with your permission) to introduce some data oriented design in this already awesome engine.

P.S(Μήπως είσαι από την ελλάδα?) :)

**Comment by PanosK92** [comment]
Hi there!

Right now, all components contain both the data (pointers all over memory, not great) and the logic. This is what most engines seem to do, but, it's old school thinking and there are a lot of benifits to leaving it behind for a more data oriented design.

Having all components linearly arranged will reduce cache misses, allow for functions to access/return them in a fast manner, and in general, if taken further, could decouple the data from the actual logic, resulting in cleaner code. Which makes life easy, which is good.

So by all means go ahead 👍 

Ναί, απο ελλάδα 😃

**Comment by GPUCode** [comment]
Thanks! I am working on it right now. If you want you can take a look at my repository to get an idea of how I am going to implement it:
 https://github.com/GPUCode/Neon-Engine
Also could you maybe tell how to commit to your repository? Event though i have programming for some time now, I only recently started using git/github and I am a noob at it

**Comment by PanosK92** [comment]
Hello,

I had a look and if I understand correctly, everything under the "Components" directory is what you are aiming to do. The core idea seems to be that you have a templated class (ComponentManager) which keeps the components in an std::array and handles them accordingly. It's nice.

What do you think about merging all component related files into one .h/.cpp file ? It's fine as it is, I was just thinking that it might allow for some further simplifications, like removing multiple definitions of "typedef unsigned int ComponentInstance". However, these are tiny details. Don't worry about it too much.

Thanks for your interest GPUCode 😃

**Comment by GPUCode** [comment]
Yes that right! I dont mind combining all the code in one file, it makes code more compact and life easier :smiley: . 
Also could you maybe give your opinion about the code style that fits better with your engine:
```
class Entity
{
       unsigned int id;
};

class EntityImpl / EntityRef
...
private:
        Entiry e;

```
or
```
class Entity
...
private:
       unsigned int id;
``` 
Additionally would it be a hassle for you to give me a list of files that use your entity class so i know where to modify the code accordingly?

**Comment by PanosK92** [comment]
There should be a document with some coding convetion guidelines but I completely forgot about it. I will take care of it in the comming days. However, in general, I try to follow the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle), so I would go for the second example, as it has one less class and allows you to see (directly) that the Entity class boils down to an id.

You should be able to intergrate your code by tweaking the Entity class. However, if you need to find references, for any code element, and you are using Visual Studio, you can do it like [that](https://docs.microsoft.com/en-us/visualstudio/ide/finding-references?view=vs-2019).

**Comment by GPUCode** [comment]
Hello, after an eternity i have a partial implementation working. It was a bit more difficult then i expected becuase of that single component file and circular dependecies but now it works without throwing exceptions all the time. Do you mind debugging it properly?

**Comment by PanosK92** [comment]
Excellent. 

If you do a pull request I should be able to have a look at the code and also run it locally.

**Comment by GPUCode** [comment]
Actually before I do a pull request I have an important dilema that I have to sort out. Your current engine structure seems to be (if I understood correctly) that each frame each entity gets updated on it own and updates all of it's components. Now this intorduces a problem. Because now the components are stored together in an array externally, in order to keep this existing structure i have to gather them in a local vector and then perform the update. What I could also do is change this existing structure in a way that each frame ex. all entities with a T component get updated like:
```
UpdateAllEntitiesWithTransform();
UpdateAllEntitiesWithRigidBody();
```
But this solution I am afraid might introduce dependecy problems but I am not totally sure.
Also sorry that I cant do a pull request but this afternoon it broke again and now a material is nullptr and throwing exceptions :cry:

**Comment by PanosK92** [comment]
The entire world update happens at [this](https://github.com/PanosK92/SpartanEngine/blob/6d8e1820a4a51fd06690aac20bc3e0b22fefff4b/Runtime/World/World.cpp#L115) line. So if you have all the components in an external array somewhere, then only the World would need to "talk" with your system and make it update everything once, just before we actually tick any entities

I believe that this would later also allow for the removal of the components ticking by the [entity](https://github.com/PanosK92/SpartanEngine/blob/6d8e1820a4a51fd06690aac20bc3e0b22fefff4b/Runtime/World/Entity.cpp#L150) itself (the per frame operation you mentioned), and if I am not mistaken we should be on the right path, as per the data oriented philoshopy, objects like that should act only as data containers, on which other systems operate.

You can do a pull request whenever you want, and the design doesn't have to be perfect, we'll have a look at it together whenever you feel like it. So no worries and no pressure, take your time 😃

**Comment by PanosK92** [comment]
By the way, regarding the following:
```
UpdateAllEntitiesWithTransform();
UpdateAllEntitiesWithRigidBody();
```
Do you think it's possible to turn them into something more abstract, like that ?
```
UpdateAllEntities<T>()
```
I think it might make the code easier to maintain, for example implenting new components will be easier as it will require less manual work (e.g. the coder won't have to go through many files to update/add stuff).

However, these are just details that can be tweaked at any moment. Get your idea to work however you think is best.

**Comment by GPUCode** [comment]
It should not be that difficult to implement a template function like that. But now I have more problems to face as I dig deeper into your engine:
 1. How to update the components and entites directly from the world.
2. How to handle disabled entities and serialization
3. How to iterate over the templated managers and perform the update

For problem 1 I think a good solution might be (pseudocode):
```
for (manager in managers)
{
     for (component in manager)
     {          
            component->Tick();
     }
}
```

To solve the first part of problem 2 I just need to add a condition before updating:
```
for ( manager in managers)
{
     for (component in manager)
     {          
            if (component->IsParentEntityActive())
                     component->Tick();
     }
}
```  

For the second part of problem 2 I cant do anything because I have not studied how to do serialazation and help is needed!

Now all of these problems boil down to iterating over the managers which are templated classes
Currently they all inherit from a base class called BaseComponentManager which is not templated and so it cant have virtual declarations of the function the manager has becuase these functions are templated. So a static cast is required to cast a BaseComponentManager to a ComponentManager<T>. The problem is finding out which manager is which.

The first solution i tried was this:
```
//Your enum which indicates which component is begin managed as key and and manager itself
std::map<ComponentType, BaseComponentManager*>

void IterateManagers(std::function<void(BaseComponentManager*, ComponentType)> func);
...
IterateManagers([&](BaseComponentManager* manager, ComponentType type)
{
        switch(type)
        {
        case ComponentType_Transform:
        {
              auto m = static_cast<ComponentManager<Transform>(manager);
              m->GetComponent(GetId()); 
        }
        ...
        }
});
``` 

(I ditched the EntityID because the Enitity inherites from Spartan_Object which already provides an id so why not use that?)

But this solution is not only rigid and requires a ton of code for only a simple thing it is also a lot harder to maintain and add new components. So with the help of our friend the tuple, some template madness and stackoverflow I pieced together this solution:

Iteration.h
```
namespace Spartan
{
    template<std::size_t I = 0, typename FuncT, typename... Tp>
    inline typename std::enable_if<I == sizeof...(Tp), void>::type
        for_all(const std::tuple<Tp...>&, FuncT) // Unused arguments are given no names.
    { }

    template<std::size_t I = 0, typename FuncT, typename... Tp>
    inline typename std::enable_if < I < sizeof...(Tp), void>::type
        for_all(const std::tuple<Tp...>& t, FuncT f)
    {
        f(std::get<I>(t));
        for_all<I + 1, FuncT, Tp...>(t, f);
    }

    template<std::size_t I = 0, typename FuncT, typename... Tp>
    inline typename std::enable_if<I == sizeof...(Tp), void>::type
        for_all(std::tuple<Tp...>&&, FuncT) // Unused arguments are given no names.
    { }

    template<std::size_t I = 0, typename FuncT, typename... Tp>
    inline typename std::enable_if < I < sizeof...(Tp), void>::type
        for_all(std::tuple<Tp...>&& t, FuncT f)
    {
        f(std::get<I>(t));
        for_all<I + 1, FuncT, Tp...>(std::move(t), f);
    }
}
``` 

World.h
```
...
        //= COMPONENT MANAGERS ========================
        template <typename T>
        auto GetComponentManager();

        template <typename Func>
        void IterateManagers(Func f);
        //================================================

...

template<typename T>
    auto World::GetComponentManager()
    {
        ComponentType type = IComponent::TypeToEnum<T>();

        BaseComponentManager* baseManager = m_component_managers[type].get();

        return static_cast<ComponentManager<T>*>(baseManager);
    }

    template<typename Func>
    inline void World::IterateManagers(Func f)
    {
        for_all(std::tie(Transform(), RigidBody(), Light(), Collider(),
                        Constraint(), Camera(), Script(), AudioListener(),
                        AudioSource(), Skybox()), [&](auto type)
       {
           ComponentManager<decltype(type)>* manager = this->GetComponentManager<decltype(type)>();
           f(manager);
       });
    }
```

Usage:
```
m_context->GetSubsystem<World>()->IterateManagers([&](auto& manager)
{
      manager->GetComponent(GetId());
      ...
});
```

The objects in the std::tie call are all there just for decltype to get their type. We do not care about their adress or if the are NULL. So with this solution adding new components requires only to add the new object in the std::tie list.

But this requires the use of auto lambdas which are only supported on C++ 14 and above so now your codebase must support C++ 14 😢

I am really looking for your feedback about this and any suggestions to improve it.
I fixed the material problem with a few extra checks in the ComponentManager now it all works now perfectly so I think I might do a pull request tommorow if everything goes according to plan and nothing breaks 😃 

Also do you have any idea on how this would work on your double buffered actor system? Should I also double buffer the components? I am not sure

**Comment by GPUCode** [comment]
Here is the pull request: https://github.com/PanosK92/SpartanEngine/pull/28

**Comment by PanosK92** [comment]
Perfect 👍 

Thanks for doing this!

I will have a look at everything and get back to you as soon as possible (during the weekend at the latest) as this week is turning out to be quite busy 🏃
